x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/AutomatonFieldPredicate.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.permission;
+
+import org.apache.lucene.util.RamUsageEstimator;
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.CharacterRunAutomaton;
+import org.apache.lucene.util.automaton.Transition;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.io.stream.OutputStreamStreamOutput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.plugins.FieldPredicate;
+
+import java.io.IOException;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.util.Base64;
+
+/**
+ * An implementation of {@link FieldPredicate} which matches fields
+ * against an {@link Automaton}.
+ */
+class AutomatonFieldPredicate implements FieldPredicate {
+    private final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(AutomatonFieldPredicate.class);
+
+    private final String automatonHash;
+    private final CharacterRunAutomaton automaton;
+
+    AutomatonFieldPredicate(Automaton originalAutomaton, CharacterRunAutomaton automaton) {
+        this.automatonHash = sha256(originalAutomaton);
+        this.automaton = automaton;
+    }
+
+    @Override
+    public boolean test(String field) {
+        return automaton.run(field);
+    }
+
+    @Override
+    public String modifyHash(String hash) {
+        return hash + ":" + automatonHash;
```

Comment:
I thought we were going to disable mapping hash with FLS, but this is clever!

Reply:
Sometimes I can be clever. Sometimes!